src/Microsoft.Health.Dicom.Core.UnitTests/Features/Common/ParallelEnumerableTests.cs

```diff
@@ -0,0 +1,185 @@
+        await Assert.ThrowsAsync<ArgumentNullException>(() => input.SelectParallel((Func<string, CancellationToken, ValueTask<int>>)null, new ParallelEnumerationOptions()).ToListAsync().AsTask());
+        await Assert.ThrowsAsync<ArgumentNullException>(() => input.SelectParallel(ParseAsync, null).ToListAsync().AsTask());
+
+        var options = new ParallelEnumerationOptions { MaxBufferedItems = 1, MaxDegreeOfParallelism = -1 };
+        await Assert.ThrowsAsync<ArgumentException>(() => input.SelectParallel(ParseAsync, options).ToListAsync().AsTask());
+
+        options = new ParallelEnumerationOptions { MaxBufferedItems = 5, MaxDegreeOfParallelism = 6 };
+        await Assert.ThrowsAsync<ArgumentException>(() => input.SelectParallel(ParseAsync, options).ToListAsync().AsTask());
+    }
+
+    [Fact]
+    public async Task GivenNoElements_WhenSelectingInParallel_ThenReturnEmpty()
+    {
+        var input = Enumerable.Empty<string>();
+        Assert.Empty(await input.SelectParallel(ParseAsync, new ParallelEnumerationOptions()).ToListAsync());
+    }
+
+    [Fact]
+    public async Task GivenOneElement_WhenSelectingInParallel_ThenReturnOneElement()
+    {
+        var input = new List<string> { "42" };
+        Assert.Equal(42, await input.SelectParallel(ParseAsync, new ParallelEnumerationOptions()).SingleAsync());
+    }
+
+    [Fact]
+    public async Task GivenFewerThenParallelism_WhenSelectingInParallel_ThenReturnAllResults()
+    {
+        var input = new List<string> { "1", "2", "3" };
+        await AssertValuesAsync(
+            input.SelectParallel(ParseAsync, new ParallelEnumerationOptions { MaxDegreeOfParallelism = input.Count + 1 }),
+            1, 2, 3);
+    }
+
+    [Fact]
+    public async Task GivenMoreThenParallelism_WhenSelectingInParallel_ThenReturnAllResults()
+    {
+        var input = new List<string> { "1", "2", "3", "4", "5" };
+        await AssertValuesAsync(
+            input.SelectParallel(ParseAsync, new ParallelEnumerationOptions { MaxDegreeOfParallelism = input.Count - 2 }),
+            1, 2, 3, 4, 5);
+    }
+
+    [Fact]
+    public async Task GivenSource_WhenEnumeratingMultipleTimes_ThenReturnSameishResults()
+    {
+        var input = new List<string> { "1", "2", "3", "4", "5" };
+        for (int i = 0; i < 5; i++)
+        {
+            await AssertValuesAsync(
+                input.SelectParallel(ParseAsync, new ParallelEnumerationOptions { MaxDegreeOfParallelism = input.Count - 2 }),
+                1, 2, 3, 4, 5);
+        }
+    }
+
+    [Fact]
+    public async Task GivenError_WhenSelectingInParallel_ThenRethrowError()
+    {
+        var input = new List<string> { "1", "foo", "3" };
+        await Assert.ThrowsAsync<FormatException>(
+            () => input.SelectParallel(ParseAsync, new ParallelEnumerationOptions()).ToListAsync().AsTask());
+    }
+
+    [Fact]
+    public async Task GivenCancelledToken_WhenSelectingInParallel_ThenThrowException()
```

Comment:
Where will we catch this? I don't see a catch up the stack and we don't handle `TaskCanceledException` in the middleware currently.

Reply:
I've added a test case, but we already check for `OperationCanceledException`